lib/web_ui/lib/src/engine/semantics/semantics.dart

```diff
@@ -509,6 +538,23 @@ abstract class PrimaryRoleManager {
   /// the object.
   @mustCallSuper
   void update() {
+    final Map<Role, RoleManagerFactory>? secondaryRoleFactories = _secondaryRoleFactories;
+    if (secondaryRoleFactories != null && secondaryRoleFactories.isNotEmpty) {
+      secondaryRoleFactories.forEach((Role role, RoleManagerFactory roleFactory) {
+        final RoleManager? secondaryRoleManager = roleFactory(semanticsObject);
+        if (secondaryRoleManager != null) {
+          _secondaryRoleManagers ??= <RoleManager>[];
+          _secondaryRoleManagers!.add(secondaryRoleManager);
+        }
+      });
+
+      // A role manager should only be added once. So remove any factories that
+      // contributed a role manager.
+      _secondaryRoleManagers?.forEach((RoleManager roleManager) {
+        secondaryRoleFactories.remove(roleManager.role);
+      });
```

Comment:
To clarify my suggestion above, here's how I'm imagining this method to look like:
```dart
    final List<Role> secondaryRoles = _secondaryRoles;
    for (Role role in secondaryRoles) {
      final RoleManager? secondaryRoleManager = role.createManager();
      if (secondaryRoleManager != null) {
        _secondaryRoleManagers ??= <RoleManager>[];
        _secondaryRoleManagers!.add(secondaryRoleManager);
        _secondaryRoles.remove(role); // Is this allowed inside the for loop?
      }
    }
```